### PR TITLE
drop audio tags that come before buffered end

### PR DESCRIPTION
**Pull request by mjneil**
PR #106 removed the dropping of FLV tags that came before the end of the buffer before appending to prevent artifacts due to missing I frames. This created an issue that caused audio to replay when appending a repeated segment from another rendition. It appears that when we append video data that has timestamps before the playhead, flash ignores those tags and does not replay video, but seems to not take into account the timestamps of the audio and appends the audio data to the buffer, causing a stall in video playback and a repeat in audio. 

This change trims the audio and video FLV tags separately to bring back dropping of audio tags that come before the end of the buffer, but keeping the video tags. Both types are still dropped on seeks.

### CODE REVIEW

File: src/flash-source-buffer.js [hunk 1]
```diff
@@ -304,47 +304,68 @@ export default class FlashSourceBuffer extends videojs.EventTarget {
...
+    targetAudioPts *= 1e3;
+    targetAudioPts += this.basePtsOffset_;
 
     // skip tags with a presentation time less than the seek target
-    for (i = 0; i < tags.length; i++) {
-      if (tags[i].pts >= targetPts) {
-        filteredTags.push(tags[i]);
+    for (i = 0; i < videoTags.length; i++) {
+      if (videoTags[i].pts >= targetVideoPts) {
+        filteredVideoTags.push(videoTags[i]);
```
Comment (imbcmdth) [inline src/flash-source-buffer.js]: We seem to not culling any video tags at all (effectively). I think the idea should cull with a bit of awareness of the GOP structure and cull entire GOPS that precede the `targetVideoPts` (which would be the seek time or the end of the buffer). Doing this means playback starts within complete GOPS and should eliminate the "smearing" that was experienced previously without pushing too many extra frames into Flash.

File: test/flash.test.js [hunk 2]
```diff
@@ -343,6 +343,54 @@ QUnit.test('drops tags before currentTime when seeking', function() {
...
+
+  endTime = 10 + 7;
+  this.mediaSource.tech_.trigger('seeking');
+  sourceBuffer.appendBuffer(new Uint8Array(10));
+  this.swfCalls.length = 0;
+  timers.runAll();
+
+  // 0 - 9 for video tags not droped, 7 - 9 for audio tags (0 - 6 dropped for being before buffered end)
+  QUnit.deepEqual(this.swfCalls[0].arguments[0], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 7, 8, 9],
+            'three tags are appended');
```
Comment (imbcmdth) [inline test/flash.test.js]: This assertions description doesn't make sense - 13 tags are inserted!

File: test/flash.test.js [hunk 3]
```diff
@@ -343,6 +343,54 @@ QUnit.test('drops tags before currentTime when seeking', function() {
...
+  // starting PTS value
+  while (i--) {
+    tags_.unshift(
+      makeFlvTag((i * 1000) + (29 * 1000),
+        new Uint8Array([i])));
+  }
+  sourceBuffer.segmentParser_.trigger('data', {
+    tags: {
+      videoTags: tags_,
+      audioTags: tags_
```
Comment (imbcmdth) [inline test/flash.test.js]: `videoTags` and `audioTags` should have easily identifiable tags so that we can be sure that the right ones are culled. Right now you can't be sure which 0-9 and which 7-9 there are represented in the assertion below.

File: test/flash.test.js [hunk 4]
```diff
@@ -343,6 +343,54 @@ QUnit.test('drops tags before currentTime when seeking', function() {
...
+  });
+
+  endTime = 10 + 7;
+  this.mediaSource.tech_.trigger('seeking');
+  sourceBuffer.appendBuffer(new Uint8Array(10));
+  this.swfCalls.length = 0;
+  timers.runAll();
+
+  // 0 - 9 for video tags not droped, 7 - 9 for audio tags (0 - 6 dropped for being before buffered end)
+  QUnit.deepEqual(this.swfCalls[0].arguments[0], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 7, 8, 9],
```
Comment (imbcmdth) [inline test/flash.test.js]: Shouldn't the video and audio tags be interleaved by time so that the output would be: `[0, 1, 2, 3, 4, 5, 6, 7, 7, 8, 8, 9, 9]`??

File: src/flash-constants.js [hunk 5]
```diff
@@ -13,7 +13,7 @@
  */
 const flashConstants = {
   // times in milliseconds
-  TIME_BETWEEN_CHUNKS: 4,
+  TIME_BETWEEN_CHUNKS: 2,
```
Comment (imbcmdth) [inline src/flash-constants.js]: Please revert this change, please. Other than this it LGTM!

**Comment by forbesjo** [comment]
Looks good to me.
